Use `pytree.tree_leaves` everywhere

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112324
* #112323
* #112327

This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi

peterbell10 committed:
Use `pytree.tree_leaves` everywhere

This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

XuehaiPan commented on torch/distributed/_spmd/api.py:
```diff
@@ -340,7 +340,7 @@ def _compile(
     # FIXME(@mrshenli): need to broadcast model to sync parameters
     mod, opt = None, None
-    for arg in pytree.tree_flatten(list(args) + list(kwargs.values()))[0]:
+    for arg in pytree.tree_leaves(list(args) + list(kwargs.values())):
```
```suggestion
    for arg in pytree.tree_leaves((args, kwargs)):
```

lezcano replied:
kwargs.values()* Same below.

XuehaiPan commented on torch/nn/utils/_per_sample_grad.py:
```diff
@@ -62,7 +62,7 @@ def maybe_build_expanded_weight(og_tensor, batch_size):
 
     def compute_batch_size(*args, **kwargs):
-        args_and_kwargs = tree_flatten(args)[0] + tree_flatten(kwargs)[0]
+        args_and_kwargs = pytree.tree_leaves(args) + pytree.tree_leaves(kwargs)
```
```suggestion
        args_and_kwargs = pytree.tree_leaves((args, kwargs))
```

lezcano reviewed (approved):
LGTM, but it'd be good to add the optimisations from @XuehaiPan as well

peterbell10 commented:
Yeah I have some other ideas to optimize the `args, kwargs` calls but didn't want to get too clever in this PR since it's fairly large. Most of this PR was done with simple codemod regex replacements.

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
BTW I did run the benchmarks and there is a noticeable reduction in mean compile time for this stack. Only ~1-5% but not bad for a couple relatively small changes.

https://hud.pytorch.org/benchmark/compilers?startTime=Sun%2C%2022%20Oct%202023%2014%3A03%3A59%20GMT&stopTime=Sun%2C%2029%20Oct%202023%2015%3A03%3A59%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/peterbell10/644/orig&lCommit=e5bf8b172b945adf50484ae8025db45b97e228e4&rBranch=main&rCommit=0948550c53a3b6634b4fdf330d76eb8236c411c9

XuehaiPan commented:
The main branch has moved the functions in `pytree_hacks.py` to `_pytree.py`. You need to rebase commits onto the main branch.

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
Update on "Use `pytree.tree_leaves` everywhere"


This changes all the instances I could find of `tree_flatten(...)[0]` or
`x, _ = tree_flatten` to use `tree_leaves`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
